File: Help the bookseller !/help-the-bookseller.py
def stock_list(listOfArt: 'list', listOfCat: 'list') -> 'str':
    """
    :type listOfArt: list
    :type listOfCat: list
    :rtype: str
    """
    if not len(listOfArt) or not len(listOfCat):
        return ''

    newList = dict()

    for item in listOfArt:
        key, val = item.split(" ")
        newList[key[0]] = newList[key[0]] + int(val) if key[0] in newList else int(val)

    return " - ".join(list(map(lambda x: '({} : {})'.format(x, newList[x] if x in newList else 0), listOfCat)))

# (A : 0) - (B : 1290) - (C : 515) - (D : 600)
print(stock_list(["BBAR 150", "CDXE 515", "BKWR 250", "BTSQ 890", "DRTY 600"], ["A", "B", "C", "D"]))
# (A : 200) - (B : 1140)
print(stock_list(["ABAR 200", "CDXE 500", "BKWR 250", "BTSQ 890", "DRTY 600"], ["A", "B"]))

File: Help the bookseller !/help-the-bookseller.php
<?php
/**
 * Created by PhpStorm.
 * User: kovalevcon
 * Date: 19.03.19
 * Time: 18:25
 */

/**
 * Complete the stockList function in the editor below.
 *
 * @param array $listOfArt
 * @param array $listOfCat
 * @return string
 */
function stockList(array $listOfArt, array $listOfCat): string
{
    if (!count($listOfArt) || !count($listOfCat)) {
        return '';
    }

    $newList = [];

    foreach ($listOfArt as $item) {
        $arr = explode(" ", $item);
        $key = $arr[0][0];

        $newList[$key] = array_key_exists($key, $newList) ?  $newList[$key] + (int)$arr[1] : (int)$arr[1];
    }

    return implode(' - ', array_map(function ($val) use ($newList) {
        return "({$val} : " . (array_key_exists($val, $newList) ? $newList[$val] : 0) . ")";
    }, $listOfCat));
}

echo "<pre>stockList([\"BBAR 150\", \"CDXE 515\", \"BKWR 250\", \"BTSQ 890\", \"DRTY 600\"], [\"A\", \"B\", \"C\", \"D\"])\r\n";
// (A : 0) - (B : 1290) - (C : 515) - (D : 600)
echo stockList(["BBAR 150", "CDXE 515", "BKWR 250", "BTSQ 890", "DRTY 600"], ["A", "B", "C", "D"]);
echo "</pre>";

echo "<pre>stockList([\"ABAR 200\", \"CDXE 500\", \"BKWR 250\", \"BTSQ 890\", \"DRTY 600\"], [\"A\", \"B\"])\r\n";
// (A : 200) - (B : 1140)
echo stockList(["ABAR 200", "CDXE 500", "BKWR 250", "BTSQ 890", "DRTY 600"], ["A", "B"]); // 33
echo "</pre>";